File: preface/smallest_free_id/improvement_2/min_free.py
import time
import random

HUNDRED_THOUSAND = 100_000
MILLION = 1_000_000


def min_free(lst, l, u):
    m = (l + u) // 2
    if not lst:
        return l
    left = [i for i in lst if i <= m]
    right = [i for i in lst if i > m]
    if len(left) == m - l + 1:
        return min_free(right, m + 1, u)
    else:
        return min_free(left, l, m)


def min_free_(lst):
    l = 0
    u = len(lst) - 1
    while len(lst) > 0:
        m = (l + u) // 2
        left = 0
        right = 0
        for right, item in enumerate(lst):
            if item <= m:
                lst[left], lst[right] = lst[right], lst[left]
                left += 1
        if left == m - l + 1:
            lst = lst[left:]
            l = m + 1
        else:
            lst = lst[:left]
            u = m
    return l


def main():
    total = 0
    loop_times = 10
    for i in range(loop_times):
        data = random.sample(range(0, MILLION + loop_times), k=MILLION)
        start_time = time.clock()
        v1 = min_free(data, 0, len(data) - 1)
        print(f'result 1: {v1}')
        v2 = min_free_(data)
        print(f'result 2: {v2}')
        total += time.clock() - start_time
    print(f'Average CPU time: {total / loop_times}s')


if __name__ == '__main__':
    main()

File: preface/smallest_free_id/improvement_2/min_free.cpp
#include <algorithm>
#include <ctime>
#include <iostream>
#include <vector>
#include "../generate_data.cpp"
using namespace std;

int min_free(vector<int>::iterator begin, vector<int>::iterator end, int l,
             int u) {
  int m = (l + u) / 2;
  if (begin == end) {
    return l;
  }
  auto it = partition(begin, end, [m](int i) { return i <= m; });
  if (it - begin == m - l + 1) {
    return min_free(it, end, m + 1, u);
  } else {
    return min_free(begin, it, l, m);
  }
}

int min_free_(vector<int> &lst) {
  int l = 0;
  int u = lst.size() - 1;
  auto begin = lst.begin();
  auto end = lst.end();
  while (end - begin) {
    int m = (l + u) / 2;
    auto left = partition(begin, end, [m](int i) { return i <= m; });
    if (left - begin == m - l + 1) {
      begin = left;
      l = m + 1;
    } else {
      end = left;
      u = m;
    }
  }
  return l;
}

int main() {
  // int hundred_thousand = 100000;
  int million = 1000000;
  clock_t total = 0;
  size_t loop_times = 10;
  for (size_t i = 0; i < loop_times; i++) {
    auto lst = generate_data(million, loop_times);
    clock_t start_time = clock();
    int v1 = min_free(lst.begin(), lst.end(), 0, lst.size() - 1);
    cout << "result 1: " << v1 << endl;
    int v2 = min_free_(lst);
    cout << "result 2: " << v2 << endl;
    total += clock() - start_time;
  }
  cout << static_cast<double>(total) / loop_times / CLOCKS_PER_SEC << endl;
}
